tests/admin_views/test_multidb.py

```diff
@@ -0,0 +1,55 @@
+from unittest.mock import patch
+
+from django.conf.urls import url
+from django.contrib import admin
+from django.contrib.auth.models import User
+from django.db import connections, router
+from django.test import TestCase, override_settings
+
+from .models import Book
+
+
+class Router(object):
+    def __init__(self, target):
+        self.target = target
+
+    def db_for_read(self, model, **hints):
+        return self.target
+
+    def db_for_write(self, model, **hints):
+        return self.target
+
+site = admin.AdminSite(name="test_adminsite")
+site.register(Book)
+
+urlpatterns = [
+    url(r'^admin/', site.urls),
+]
+
+
+@override_settings(ROOT_URLCONF=__name__)
+class MultiDatabaseTest(TestCase):
+    multi_db = True
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.superuser = {}
+        for db in connections:
+            router.routers = [Router(db)]
+            cls.superuser[db] = User.objects.create_superuser(username='admin', password='something',
+                                                              email='test@test.org')
+
+    @patch('django.contrib.admin.options.transaction')
+    def test_save_to_each_db(self, mock):
+        for db in connections:
+            # The router must be set before we login.
+            router.routers = [Router(db)]
+            self.client.force_login(self.superuser[db])
+
+            book_name = 'Foobar: 5th edition'
+            self.client.post('/admin/admin_views/book/add/', {'name': book_name})
+            try:
+                self.assertEqual(mock.atomic.call_args_list[-1][1]['using'], db,
+                                 "The wrong `using` kwarg was passed to django.db.transaction.atomic!")
+            except IndexError:
+                self.fail("django.contrib.admin.options.transaction.atomic was never called!")
```

Comment:
Since atomic has only two possible arguments and savepoint=False is not
likely to be passed here I think it's safe to use assert_called_with.

2016-08-26 14:13 GMT-04:00 Jibodeah <notifications@github.com>:

> In tests/admin_views/test_multidb.py
> <https://github.com/django/django/pull/7143#discussion_r76463341>:
>
> > +                                                              email='test@test.org')
> > +
> > +    @patch('django.contrib.admin.options.transaction')
> > +    def test_save_to_each_db(self, mock):
> > +        for db in connections:
> > +            # The router must be set before we login.
> > +            router.routers = [Router(db)]
> > +            self.client.force_login(self.superuser[db])
> > +
> > +            book_name = 'Foobar: 5th edition'
> > +            self.client.post('/admin/admin_views/book/add/', {'name': book_name})
> > +            try:
> > +                self.assertEqual(mock.atomic.call_args_list[-1][1]['using'], db,
> > +                                 "The wrong `using` kwarg was passed to django.db.transaction.atomic!")
> > +            except IndexError:
> > +                self.fail("django.contrib.admin.options.transaction.atomic was never called!")
>
> I could, but I think that would also check that any other arguments to
> atomic (i.e. savepoint) are not provided. So I did it this way so as to
> only check the using argument.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/django/django/pull/7143/files/2ba1074407bf12bea45b96efb7a5f4fb5fbdf0ef#r76463341>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAkTV6xiceWHY115x5Xj6ozWr7vTLZEks5qjyy5gaJpZM4JrKFq>
> .
>